pkg/stream/udpproxy.go

```diff
@@ -0,0 +1,139 @@
+// Copyright © 2018 Infostellar, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stream
+
+import (
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+)
+
+type UDPProxy interface {
+	Start() error
+	io.Closer
+}
+
+type udpProxy struct {
+	recvConn net.PacketConn
+	sendConn net.Conn
+	stream   SatelliteStream
+
+	recvBuf       []byte
+	recvCloseChan chan struct{}
+
+	sendChan      chan []byte
+	sendCloseChan chan struct{}
+
+	closeWg sync.WaitGroup
+}
+
+// NewUDPProxy creates a UDPProxy that will listen for packets to send to the satellite and send back
+// received packets.
+func NewUDPProxy(recvAddr string, sendAddr string, satelliteId string) (UDPProxy, error) {
+	rc, err := net.ListenPacket("udp", recvAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	sc, err := net.Dial("udp", sendAddr)
+	if err != nil {
+		rc.Close()
+		return nil, err
+	}
+
+	sendChan := make(chan []byte)
+
+	stream, err := NewSatelliteStream(satelliteId, sendChan)
+	if err != nil {
+		rc.Close()
+		sc.Close()
+		return nil, err
+	}
+
+	p := &udpProxy{
+		recvConn:      rc,
+		sendConn:      sc,
+		stream:        stream,
+		sendChan:      sendChan,
+		recvBuf:       make([]byte, 1024*1024),
+		sendCloseChan: make(chan struct{}),
+		recvCloseChan: make(chan struct{}),
+		closeWg:       sync.WaitGroup{},
+	}
+	return p, nil
+}
+
+// Start starts the proxy, listening for packets to send to/from the satellite.
+func (p *udpProxy) Start() error {
```

Comment:
Since the proxy can't be restarted once it has been closed, I wonder if it would be better to just call `Start()` in the constructor?

Reply:
I tend to like not doing too complex things like I/O in constructors, what do you think?

Reply:
Generally speaking, Go seems to do this sort of stuff in constructors.  For example, you call `os.Open()` and it opens the filehandle and returns the stream to you, which you then call `Close()` on later.  You could change the constructor name to something like `StartProxy` to make it more clear to the user of the method what is happening.

Reply:
Usually when there is never a reason to create an instance of the struct without also performing the task of making it ready to use.

Reply:
Ok changed it to `StartUDPProxy`